### PR TITLE
WIP: Finalizing documentation for release

### PR DESCRIPTION
**Pull request by WrathfulSpatula**
This pull request includes a draft of a document on benchmarks for Qrack. By the end of this review process, the documentation should be ready for release with the library.

### CODE REVIEW

File: docs/performance/index.rst [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+The test machine has an 04WT2G Alienware motherboard with Alienware BIOS A15. Its CPU is an Intel(R) Core(TM) i7-4910MQ. Its GPU is an NVIDIA Corporation GM204M [GeForce GTX 970M]. Its operating system is Ubuntu 16.04.4 LTS. It has 24GB of 1600MHz DDR3 RAM in 8GBx2 and 4GBx2 SODIMM configuration.
+
+Heap profiling was carried out with Valgrind Massif. Heap sampling was limited but ultimately sufficient to show statistical confidence.
+
+Results
+*******
+
+We observed extremely close correspondence with theoretical complexity and RAM usage considerations for the behavior of all engine types. QEngineCPU and QEngineOCL require exponential time for a single time over N qubits. QUnit types with explicitly separated subsystems as per https://arxiv.org/abs/1710.05867 show constant time requirements for the same single gate.
```
Comment (bennbollay) [inline docs/performance/index.rst]: “Single time over N qubits” -> different word than ‘time’, possibly?  Not sure what that means.

File: docs/performance/index.rst [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+.. image:: cnot_single.png
+
+QEngineCPU and QEngineOCL can perform many identical gates in parallel across entangled subsystems for about the same cost as a single gate. To test this, we can apply parallel gates at once across the full width of a coherent array of qubits. (CNOT is a two bit gate, so (N-1)/2 gates are applied to odd numbers of qubits.) Notice in these next graphs how QEngineCPU and QEngineOCL have approximately the same scaling cost as the single gate graphs above, while QUnit types show a linear trend (appearing logarithmic on an exponential axis scale):
+
+.. image:: x_all.png
+
+.. image:: cnot_all.png
+
+Heap sampling showed high confidence adherence to theoretical expecations. Complex numbers are represented as 2 double (64-bit) accuracy floating point types, for real and imaginary components. There is one complex number per permutation in a separable subsystem of qubits. QUnit explicitly separates subsystems, while QEngine maintains complex amplitudes for all 2^N permutations of N qubits. QEngines duplicate their state vectors once for speed and simplicity where it eases implementation.
```
Comment (bennbollay) [inline docs/performance/index.rst]: 2^N needs to be LaTex’ed.

File: docs/performance/index.rst [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+Results
+*******
+
+We observed extremely close correspondence with theoretical complexity and RAM usage considerations for the behavior of all engine types. QEngineCPU and QEngineOCL require exponential time for a single time over N qubits. QUnit types with explicitly separated subsystems as per https://arxiv.org/abs/1710.05867 show constant time requirements for the same single gate.
+
+.. image:: x_single.png
+
+.. image:: cnot_single.png
+
+QEngineCPU and QEngineOCL can perform many identical gates in parallel across entangled subsystems for about the same cost as a single gate. To test this, we can apply parallel gates at once across the full width of a coherent array of qubits. (CNOT is a two bit gate, so (N-1)/2 gates are applied to odd numbers of qubits.) Notice in these next graphs how QEngineCPU and QEngineOCL have approximately the same scaling cost as the single gate graphs above, while QUnit types show a linear trend (appearing logarithmic on an exponential axis scale):
```
Comment (bennbollay) [inline docs/performance/index.rst]: (N-1)/2 needs to be LaTex’ed

File: docs/performance/index.rst [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+  the desired target system for accurate performance assessments.
+
+* Benchmarking is Hard - While we've attempted to perform clean and accurate
+  results, bugs and mistakes do occur.  If flaws in process are identified,
+  please let us know!
+
+Method
+******
+
+100 timed trials of each method were run for each qubit count between 3 and 24 qubits. The average and quartile boundary values of each set of 100 were recorded and graphed. Grover's search to invert a black box subroutine, or "oracle," was similarly implemented for trials between 3 and 17 qubits. Grover's algorithm was iterated an optimal number of times, vs. qubit count, to maximize probability on a half cycle of the algorithm's period, being :math:`floor(π/{4*asin[sqrt(2^N)]})` iterations.
```
Comment (bennbollay) [inline docs/performance/index.rst]: Sqrt needs to be latex’ed.

File: docs/performance/index.rst [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+.. image:: grovers.png
+
+A representative sample of Qrack methods were run for 100 trials per qubit as above, for parallel gates up to the full span of the qubits. Multiple bit gates spanned the full length of coherent qubits up to integer division flooring for 2 and 3 qubit gates. Taking an observed threshold of 10 to 15 qubits for API method overhead to become much larger than "noise" levels, we regressed the high qubit end of each graph for an exponential fit for time against qubits. These regression equations are presented in tables of representative samplings of the API. The results follow this equation:
+
+.. math::
+   :label: regression_eq
+
+   [Milliseconds] = \exp \left( [Base] \left( [No. of Qubits] + [Intercept] \right) \right)
+
```
Comment (bennbollay) [inline docs/performance/index.rst]: R^2 needs to be LaTex’ed

File: docs/performance/index.rst [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+Software
+========
+
+These are a representative sample of regression equations for QEngineCPU. Testing was carried out on parallel gates across the full width of a coherent unit of quantum memory, up to integer flooring on 2 and 3 qubit gates.
+
+.. csv-table:: Regressed QEngineCPU Speed Equations
+  :header: "Method","First Qubit","Base","Intercept","R^2","p-value"
+  :widths: auto
+  
+  "INC",12,0.815,-18.97,0.996,8.7E-15
```
Comment (bennbollay) [inline docs/performance/index.rst]: CSV data should be alphabetically sorted for both tables

File: docs/performance/index.rst [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+  "X",16,0.933,-16.2,0.991,1.88E-08
+  "ROL",15,0.856,-15.67,0.993,6.02E-10
+  "PhaseFlip",13,0.646,-15.47,0.993,3.13E-12
+  "M",12,0.62,-15.2,0.993,4.56E-13
```
Comment (bennbollay) [inline docs/performance/index.rst]: SigDigit inconsistencies between values.

File: docs/performance/index.rst [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+We will also develop and maintain systematic comparisons to published benchmarks of quantum computer simulation standard libraries, as they arise.
+
+Conclusion
+**********
+
+Per `arXiv:1710.05867`_, explicitly separated subsystems of qubits in QUnit have a significant RAM and speed edge in many cases over the "Schrödinger algorithm" of QEngineCPU and QEngineOCL. One of Qrack's greatest new optimizations to either general algorithm is constant complexity or "free" scaling of bitwise parallelism in entangled subsystems, compared to linear complexity scaling without this optimization. Qrack gives at least reasonably efficient performance on a single node up to approximately 30 qubits, in the limit of maximal entanglement.
```
Comment (bennbollay) [inline docs/performance/index.rst]: Use the `Breaking the 49-Qubit...` full description to leverage the citation correctly.  It’s kinda dumb but seems to be the way sphinx docs want it to be done.

File: docs/performance/index.rst [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+
+Citations
+*********
+
+.. target-notes::
+
+.. _`Q#`: http://TBD.com
```
Comment (bennbollay) [inline docs/performance/index.rst]: https://www.microsoft.com/en-us/quantum/development-kit

File: docs/performance/index.rst [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+.. _`QHipster`: https://TBD.com
```
Comment (bennbollay) [inline docs/performance/index.rst]: https://github.com/intel/Intel-QS

File: docs/performance/index.rst [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+.. _`Quantiki`: https://www.quantiki.org/wiki/list-qc-simulators
+.. _`Breaking the 49-Qubit Barrier in the Simulation of Quantum Circuits`: https://arxiv.org/abs/1710.05867
```
Comment (bennbollay) [inline docs/performance/index.rst]: What happened to the grover paper?  Did that move somewhere else?

File: docs/theory.rst [hunk 12]
```diff
@@ -23,6 +23,7 @@ Grover Search Algorithm
 .. [Grover] `Grover Search Algorithm <https://en.wikipedia.org/wiki/Grover%27s_algorithm>`_
 .. [GroverSummary] `Introduction to Implementing Grover's Search Algorithm <http://twistedoakstudios.com/blog/Post2644_grovers-quantum-search-algorithm>`_
 .. [GroverVisual] `Visualization of Grover's Search Algorithm <http://davidbkemp.github.io/animated-qubits/grover.html>`_
+.. [Broda2016] `Broda, Bogusław. "Quantum search of a real unstructured database." The European Physical Journal Plus 131.2 (2016): 38. <https://arxiv.org/abs/1502.04943>`_
```
Comment (bennbollay) [inline docs/theory.rst:26]: The “theory” doc is a dead doc, it’s not listed in any of the ToC. The idea was that all of the content there has been moved to other places within the doc; the contents here should probably move to a bibliography page that’s linked to in the ToC in index.rst.

Additionally, the performance page needs to be added to the index.rst ToC as well.

File: docs/performance/index.rst (hunk 11 above)
Comment (WrathfulSpatula) [reply docs/performance/index.rst]: It was moved into its appropriate place in theory.rst.

File: docs/performance/index.rst (hunk 1 above)
Comment (WrathfulSpatula) [reply docs/performance/index.rst]: "Single gate" rather than "single time." Updating.

**Pull request merged**